sound/soc/sof/intel/hda-dai.c

```diff
@@ -309,6 +331,23 @@ static int __maybe_unused hda_dai_trigger(struct snd_pcm_substream *substream, i
 			dev_err(sdev->dev, "%s: failed to clean up link DMA\n", __func__);
 			return ret;
 		}
+
+		/*
+		 * The firmware for IPC4 expects the linkDMA to be reset but not
+		 * released on stop, the same channel must remain to be used if
+		 * we have a start after the stop without a close.
+		 * The firmware keeps it's side of the LinkDMA channel on stop,
+		 * we need to do the same on the host side.
+		 *
+		 * If the stream is closed the linkDMA is going to be cleaned up
+		 * in via hda_dai_hw_free()
+		 */
+		if (cmd == SNDRV_PCM_TRIGGER_STOP &&
+		    sdev->pdata->ipc_type == SOF_IPC_TYPE_4) {
+			hext_stream = hda_link_dma_setup(sdev, substream, dai);
+			if (IS_ERR(hext_stream))
+				return PTR_ERR(hext_stream);
+		}
```

Comment:
OTOH, this does seem correct and should work in all cases. It just looks a bit weird to first clean and then immediately setup again.

Reply:
we could add a flag parameter for hda_link_dma_cleanup() (reclaim_channel?) and hide it there? I'm not sure if that is better or worse